### PR TITLE
Reduce number of connections per node depending on the nodes role

### PR DESCRIPTION
**Pull request by s1monw**
We currently treat every node equally when we establish connections to a node.
Yet, if we are not master eligible or can't hold any data there is no point in creating
a dedicated connection for sending the cluster state or running remote recoveries respectively.
This change ensures that in these cases STATE and RECOVERY connections are shared with REGular
connections since they are unnecessary.

### CODE REVIEW

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by s1monw** [comment]
@jasontedor can you take a look

File: core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java [hunk 1]
```diff
@@ -38,11 +38,12 @@
      */
     public static final ConnectionProfile LIGHT_PROFILE = new ConnectionProfile(
         Collections.singletonList(new ConnectionTypeHandle(0, 1,
+            EnumSet.copyOf(Arrays.asList(
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java]: s/`EnumSet.copyOf(Arrays.asList(`/`EnumSet.of(`/ ?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 2]
```diff
@@ -195,20 +191,36 @@ public TcpTransport(String transportName, Settings settings, ThreadPool threadPo
...
-        builder.addConnections(connectionsPerNodeRecovery, TransportRequestOptions.Type.RECOVERY);
-        builder.addConnections(connectionsPerNodeReg, TransportRequestOptions.Type.REG);
-        builder.addConnections(connectionsPerNodeState, TransportRequestOptions.Type.STATE);
-        defaultConnectionProfile = builder.build();
+        Set<TransportRequestOptions.Type> regSet = EnumSet.of(TransportRequestOptions.Type.REG);
+        if (DiscoveryNode.isMasterNode(settings)) {
+            builder.addConnections(connectionsPerNodeState, TransportRequestOptions.Type.STATE);
+        } else {
+            // if we are not master eligible we don't need a dedicated channel to publish the state
+            regSet.add(TransportRequestOptions.Type.STATE);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: The only thing that I do not like about this is how it silently ignores mismatches in the values of `connectionsPerNodeReg` and `connectionsPerNodeState`.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 3]
```diff
@@ -195,20 +191,36 @@ public TcpTransport(String transportName, Settings settings, ThreadPool threadPo
...
+        }
+        if (DiscoveryNode.isDataNode(settings)) {
+            builder.addConnections(connectionsPerNodeRecovery, TransportRequestOptions.Type.RECOVERY);
+        } else {
+            // if we are not a data-node we don't need any dedicated channels for recovery
+            regSet.add(TransportRequestOptions.Type.RECOVERY);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Same comment here for `connectionsPerNodeReg` and `connectionsPerNodeRecovery`.

**Review by jasontedor — COMMENTED** [review]
I left two related comments, not requests for a change, just something to think about.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: you mean in the logging statement?

**Comment by s1monw** [comment]
@jasontedor I pushed a new commit

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I discussed with @s1monw what I meant by this offline and he pushed ef06a14921a250eb4b75e57b2aad931be757798d which is what I was looking for.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**